### ISSUE TITLE
Custom recipes are missing some ore dicted materials

### ISSUE THREAD

**Issue by dshadowwolf** [MachineMuse/MachineMusePowersuits#761]
With the currently available alpha (alpha-6) and using custom recipes ported from a private 1.7 pack some of the components (at this point Electrum Ingots and Coal Dust are the only one showing this) that have multiple providers (in this case, Base Metals and Thermal Foundation for Electrum and Base Minerals and EnderIO for Coal Dust) only one of them shows as viable, despite both being in the ore dictionary as "ingotElectrum".

I have run into an issue similar to this before, where a system has custom, loaded from disk recipes. In that case it was a caching issue, where the ore dictionary values were evaluated during preInit and never updated. In this case I'm not positive what a good solution would be, as adding the recipes outside of preInit is not a good idea, but it might be best to re-evaluate the cache at some later point - perhaps during the Init phase - so that mods that have been put through preInit after Modular Powersuits can have their materials with the proper ore dict's also available in the recipe.

A simple solution would be to request loading last/near last. However... that does lead to some issues, as I can think of at least one mod that also does similar.

**Comment by lehjr** [comment]
Have you at least verified the recipes? The thing is, I really can't do much to change the recipe handler, but I do know that invalid recipes likely won't load, and it should at least give you some debugging output. Bottom line is I cannot even begin to support your recipes or figure out why they don't work, sight unseen. And the recipes already load postInit because Oredict doesn't appear to be loaded early enough do do anything preInit:  https://github.com/MachineMuse/MachineMusePowersuits/blob/1.10.2-Java/src/main/java/net/machinemuse/numina/common/Numina.java#L62

**Comment by lehjr** [comment]
recipes are synced server to client here: https://github.com/MachineMuse/MachineMusePowersuits/blob/1.10.2-Java/src/main/java/net/machinemuse/numina/event/NuminaPlayerTracker.java

That reminds me, now that Thermal Expansion is a thing again, I'll need to update the recipes.

**Comment by dshadowwolf** [comment]
I have. The recipe works - with Base Metals components only where things are using the Ore Dictionary.

```json
	{
		"ingredients" : [
			[ { "oredictName" : "dustCoal" }, { "oredictName" : "slimeballBlue" }, { "oredictName" : "dustCoal" } ],
			[ { "oredictName" : "dustCoal" }, { "oredictName" : "slimeballBlue" }, { "oredictName" : "dustCoal" } ],
			[ { "oredictName" : "dustCoal" }, { "oredictName" : "slimeballBlue" }, { "oredictName" : "dustCoal" } ]
		],
		"result" : {
			"oredictName" : "componentCarbonMyofiber",
			"quantity" : 4
		}
	},
```
That is one of the recipes.  Another is:
```json
	{
		"ingredients" : [
			[ null, { "oredictName" : "componentWiring" }, null ],
			[ { "oredictName" : "componentWiring" }, { "oredictName" : "ingotElectrum" }, null ],
			[ null, { "oredictName" : "ingotElectrum" }, { "oredictName" : "componentWiring" } ]
		],
		"result" : {
			"registryName" : "powersuits:powerTool"
		}
	},
```

Both work, when the Coal Dust is the one from Base Metals and the Electrum is from the same. Thermal Foundation has its Electrum in as "ingotElectrum" and EnderIO uses 'dustCoal'.  Looking in JEI and doing testing, the Base Metals electrum ingot has 25 pages of crafting recipes - three of them specifically the custom recipes I've configured similar to the above examples - while the Thermal Foundation electrum only has 22. I've assumed that there was ore-dict resolution going on in preInit, as that is what has happened in the only other cases I've seen with similar results.

Anyway... it does seem very, very odd that I'm seeing these results. Even more odd considering that I had just updated to alpha-6 and taken the time to go through and port the recipes (though I'm sad that Thermal has moved away from registering buckets of its liquids with the ore dictionary, causing issues with using them in recipes, since they are now just the universal bucket with a fill). Looking at the code you've quoted I'm puzzled that I'm running into this at all... Unless Thermal Foundation is doing something that King Lemming recently ranted about and registering stuff in the OreDict during Post...

As for those recipes... No more multiple machine frames and energy cell frames and the coils, at least, are in Thermal Foundation now. (With meta values up in the 500's)

**Comment by lehjr** [comment]
Oredict seems to load somewhere around the init phase, or at least that was what I found while debugging a 1.10.2 port of B0bGary's Growable Ores that I was work on. I used the event handler for the ore registry to see when it was firing, and it's late enough that the automatic setup he was using for using registered ores no longer works. This is because the blocks and items need to be registered in preInt so that they aren't null when setting up the rendering for them. 

As for the bucket thing, I was just looking at that. I'll have to dig into it, but I'm pretty sure the bucket can be used. I just have to look up the syntax for it.

**Comment by lehjr** [comment]
The bucket should look something like 
```{ "registryName" : "forge:bucketFilled", "nbtString" : {"FluidName" : "cryotheum", "Amount" : 1000}}```

Edit: Sorry,  no, not quite right, this should be it: 
```{ "registryName" : "forge:bucketFilled", "nbtString" : "{FluidName:cryotheum, Amount:1000}"}```

**Comment by dshadowwolf** [comment]
oooh, thank you! I had changed the recipes to avoid the buckets, not knowing of the "nbtString" bit...

I really do prefer having the recipes as close to what was in the 1.7.10 version of my personal pack as possible... Makes life easier for me and the two or three friends that use it :)

**Comment by lehjr** [comment]
I'm working on the Thermal expansion recipes now, but probably won't be done until tomorrow. CraftTweaker helps with looking up the oredict and registry names.

**Comment by dshadowwolf** [comment]
Yeah, it is what I've almost always used - doesn't hurt that I've actually talked with Jared on Discord, either :)

**Comment by lehjr** [comment]
looks like a lot of stuff is registered under thermalfoundation rather than thermalexpansion

**Comment by dshadowwolf** [comment]
I'm doing a bit of debugging trying to track down the cause of a bug that seems like a cross between a timing bug and an xmod that has led me into dumping the ore dictionary in my own way. This is proving quite fun and frustrating at the same time. I could forward you a cleaned-up version that ties the actual ore dictionary name to the display name, if you want. (Will cover several mods, though I could trim it back to just Thermal)

**Comment by lehjr** [comment]
The best place to start would be the recipes, that and probably just a list of the mods that you're trying to support, if not the mods themselves.

**Comment by dshadowwolf** [comment]
Oddly enough... it looks to be a strange timing issue - I've dumped the recipes and the ore dictionary to get a look at them and the data. This one is more because I'm on the Base Metals and Base Minerals dev team(s) and the crash implicates Base Metals, even if we can't figure out wtf is actually going on to get MC into the state it is with the crash.

But this is getting beyond the matter now... I don't see where Numina/Powersuits would be doing anything wrong and can't see how a recipe-file entry of "oreDictionary" : "ingotElectrum" or "oreDictionary" : "dustCoal" is only finding one item...

(and yes, I'm missing the metadata from the dump - I forgot to add it when doing this up for this debugging information...)
```
[17:43:02] [Server thread/FATAL] [basemetals/basemetals]: Entries for Ore Dictionary Name ingotElectrum
[17:43:02] [Server thread/FATAL] [basemetals/basemetals]: basemetals:electrum_ingot - Electrum Ingot
[17:43:02] [Server thread/FATAL] [basemetals/basemetals]: thermalfoundation:material - Electrum Ingot
```

and...

```
[17:43:02] [Server thread/FATAL] [basemetals/basemetals]: Entries for Ore Dictionary Name dustCoal
[17:43:02] [Server thread/FATAL] [basemetals/basemetals]: basemetals:coal_powder - Coal Dust
[17:43:02] [Server thread/FATAL] [basemetals/basemetals]: ic2:dust - Coal Dust
[17:43:02] [Server thread/FATAL] [basemetals/basemetals]: thermalfoundation:material - Pulverized Coal
[17:43:02] [Server thread/FATAL] [basemetals/basemetals]: actuallyadditions:itemDust - Crushed Coal
```

This is missing the entry for EnderIO, as its from a test environment that replicates the bug/crash I mentioned, but...

**Comment by lehjr** [comment]
As far as I can see, the oredict values are working. For instance, with the new (not yet included uploaded in an alpha)Thermal Expansion recipes I have here, the ore substitutions are working fine. https://github.com/MachineMuse/MachineMusePowersuits/blob/1.10.2-Java/src/main/resources/mps-thermalexpansion.recipes
Keep in mind that in the recipes, oredict entries can show up as anything that matches that oredict value. 
Like I said, it would be easier to see the recipes themselves to see what is failing and where.

**Comment by dshadowwolf** [comment]
One of the failing recipes is included above, but if you like, I'll give you a link to the entire pack I have been playing and using for testing some mods I've either been working on or know the developer of.

**Comment by lehjr** [comment]
probably the entire pack including the recipes would be better. Then I can get a better idea of what is going on.

**Comment by dshadowwolf** [comment]
Actually tried my custom recipes (like the one above) with Thermal electrum and it didn't work - and JEI explicitely said "nope" for the auto-fill on it. 

And... lets see... where is that window with my dropbox tab open...

https://www.dropbox.com/s/z1ukmk01rgr38lj/HTW-1.10-1.0.zip?dl=0

That is the current version with the current recipes. Haven't actually checked since fixing the bits involving the buckets, but when I replaced them with other things it didn't seem to matter.

**Comment by lehjr** [comment]
I don't see any custom recipes for MPS, just the stock EnderIO ones.

**Comment by lehjr** [comment]
OK, I see the recipe issues, but not sure what the problem is yet. Took me awhile to get the dependencies and then had to update one of the mods. Anyway, looks like  Base Metals has a number of issues that might need to be addressed. I'll look into it more tomorrow.

**Comment by lehjr** [comment]
So you've given me a half of a pack where I have to hunt down the dependencies, then I look and the only recipes that I find are for EnderIO, ones I already have for supporting a mod that I do not even know whether or not you are even using since it isn't even in there. How can you possibly expect me to troubleshoot the issue with what you have given me? The whole point of this was for you to show me what you have that is failing and you have failed to do that while I have gone out of my way to try to help you. Well, that's it. I'm done.  

The only relevant issue that I can see with ModularPowersuits here is that the recipe shows at all, which is because the validation check isn't finished. So I'll go ahead and finish that and then invalid recipes won't show at all. I suggest using CratTweaker to add them.

**Comment by dshadowwolf** [comment]
woah, really ? That's... actually the entire thing I use - via Magic Launcher. And... Dammit, maybe something did get screwed up in generating that file and uploading it.

*grumbles* If you want, I'll copy my current into a curse instance and send it your way.

(though what I linked to *SHOULD* be the current - I'd just packaged it up the day you gave me a way to use the liquifacted coal and cryotheum buckets and uploaded it...)

fffff

Just noticed I'd given you the wrong link entirely

https://www.dropbox.com/s/mt94nu9sjkh1kfn/htw-mc1.10.zip?dl=0 

That is the current (well, almost, I'm testing some new mod versions right now) pack I'm testing most of my stuff in

**Comment by lehjr** [comment]
There, hopefully this was what I needed.

**Comment by lehjr** [comment]
Just went through all the recipes you have and as far as I can see they're all working so I don't know what it is I'm supposed to be seeing. The next alpha will have better recipe validation and will actually give you logging output regarding invalid recipes. Aside from that, not sure what else to do here. Oh, and a little heads up regarding custom recipes, usage of unlocalized names was removed since it's really a bad idea, especially with oredict and registry entries available. For now, I'll leave this open.

**Comment by dshadowwolf** [comment]
Hrm... if they're working for you then perhaps I need to nuke the server entirely - reset the world, the configs and everything.

Let me see if they're working SSP - if they are then it might be an issue with MPS being added to an already existing world or some other strange facet of how things are going.

Okay, so did some testing and part of it was that, for whatever reason, the one recipe was broken completely when I first made the report. The rest was that JEI is only displaying a single item as matching instead of either cycling through the different matches or displaying them as separate recipes for the same thing. I... think that might be because we've kept JEI a bit... old for API reasons in BaseMetals and my pack. I'm feeling kinda stupid for making this report now.
